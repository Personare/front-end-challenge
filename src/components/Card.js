import React, { PropTypes } from 'react';
import TweenLite from 'gsap';

export class Card extends React.Component {
    componentWillReceiveProps(nextProps) {
        if (nextProps.selected !== this.props.selected) {
            TweenLite.to(this.refs.cardFaces, 1, { rotationY: !nextProps.selected ? -180 : 0 });
        }
    }

    render() {
        return (
            <article className="card" onClick={this.props.onClick}>
                <div className="card-faces" ref="cardFaces">
                    <div className="front">
                        <img src={this.props.frontImg} alt={this.props.name} />
                    </div>
                    <div className="back">
                        <img src={this.props.backImg} alt={this.props.name} />
                    </div>
                </div>
            </article>
        );
    }
}

Card.propTypes = {
    name: PropTypes.string,
    frontImg: PropTypes.string,
    backImg: PropTypes.string,
    selected: PropTypes.bool,
    onClick: PropTypes.func,
};
