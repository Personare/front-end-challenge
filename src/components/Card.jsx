import React from "react";

const Card = ({ name = "CartjÃªnha", image = "", ...rest }) => {
  return (
    <div className="card">
      <img src={image} alt={name} />
    </div>
  );
};

export default Card;
