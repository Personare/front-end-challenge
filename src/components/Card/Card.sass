.CardContainer
  position: relative
  height: 210px
  cursor: pointer
  margin: 10px;
  .CardContent
    transform-style: preserve-3d

  .CardImage, .BackImage
    position: absolute
    width: 100px
    height: 210px
    overflow: hidden
    border-radius: 5px
    backface-visibility: hidden
    > img
      width: 100px
      height: 210px

  .BackImage
    transform: rotateY(-180deg)
