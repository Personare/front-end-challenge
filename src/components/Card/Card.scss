.Card {
    position: relative;
    perspective: 150rem;
    width: 200px;
    height: 200px;

    &__side {
        position: absolute;
        top: 0;
        left: 0;
        width: 100%;
        height: 50px;
        backface-visibility: hidden;
        transition: all .8s ease;

        &.--front,
        &.--back {
            svg {
                position: absolute;
                top: 0;
                right: 0;
                bottom: 0;
                left: 0;
                width: 1em;
                height: 1em;
                font-size: 5.4rem;
            }
        }

        &.--front {
            background-color: red;

            svg {
                fill: black;
            }
        }

        &.--back {
            background-color: blue;
            transform: rotateY(180deg);

            svg {
                fill: red;
            }
        }
    }

    &.--closed &__side.\--front {
        transform: rotateY(-180deg);
    }

    &.--closed &__side.\--back {
        transform: rotateY(0);
    }
}

