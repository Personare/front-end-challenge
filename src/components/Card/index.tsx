import { useContext } from "react";
import ReactCardFlip from "react-card-flip";
import { ICard } from "../../config/interfaces";
import GlobalContext from "../../global/GlobalContext";

interface ICardItem {
  card: ICard;
  onClick?: (name: string) => void;
}

const Card: React.FC<ICardItem> = ({ card, onClick }) => {
  const { name, image, show } = card;

  const {
    state: { data },
  }: any = useContext(GlobalContext);

  const isFlipped = !show;

  console.log(onClick);

  return (
    <div data-cy="card" onClick={() => onClick && onClick(name)}>
      <ReactCardFlip isFlipped={isFlipped} flipDirection="horizontal">
        <img
          alt={name}
          src={`${data.imagesUrl}${image}`}
          data-cy={!isFlipped && "card-up"}
        />

        <img
          alt={name}
          src={data.imageBackCard}
          data-cy={isFlipped && "card-down"}
        />
      </ReactCardFlip>
    </div>
  );
};

export default Card;
