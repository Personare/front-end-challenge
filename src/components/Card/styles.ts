import styled, { css } from "styled-components";

interface ImageProps {
  isMinimal?: boolean;
}
export const Image = styled.img<ImageProps>`
  ${({ isMinimal }) => css`
    height: ${isMinimal ? "150px" : "300px"};
    width: ${isMinimal ? "79px" : "auto"};
    margin-bottom: ${isMinimal ? "10px" : "auto"};
  `}
`;

export const Container = styled.div``;
