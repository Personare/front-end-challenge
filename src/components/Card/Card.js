import React, { Component } from 'react';
import TweenMax from 'gsap';

import './Card.sass';

export class Card extends Component {
  constructor(props) {
    super(props);
    this.cardContentRef = React.createRef();
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.selected !== this.props.selected) {
      TweenMax.to(this.cardContentRef.current, 1, { rotationY: !nextProps.selected ? -180 : 0 });
    }
  }

  render() {
    return (
      <article className="CardContainer" onClick={this.props.click}>
        <div className="CardContent" ref={this.cardContentRef}>
          <div className="CardImage">
            <img src={this.props.card} alt={this.props.name} />
          </div>
          <div className="BackImage">
            <img src={this.props.imageBack} alt={this.props.name} />
          </div>
        </div>
      </article>
    );
  }
}

export default Card;
