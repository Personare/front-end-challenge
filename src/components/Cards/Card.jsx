import React, { Component } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import styles from './Card.module.css'
import CardName from './CardName'
import CardImage from '../../containers/Cards/CardImage'

class Card extends Component {
  handleCardClick = () => {
    const { isShuffle } = this.props

    if (isShuffle) {
      this.setValidCurrentCard()
    }
  }

  setValidCurrentCard() {
    const { currentCard, setCurrentCard, item } = this.props

    if (!currentCard.id) {
      setCurrentCard(item)
    }
  }

  render() {
    const { isShuffle, item } = this.props

    return (
      <div
        className={classNames(styles.card, {
          [styles.backcard]: isShuffle,
        })}
        onClick={this.handleCardClick}
      >
        <CardImage isShuffle={isShuffle} item={item} />
        <CardName isShuffle={isShuffle} text={item.name} />
      </div>
    )
  }
}

Card.propTypes = {
  isShuffle: PropTypes.bool.isRequired,
  setCurrentCard: PropTypes.func.isRequired,
  item: PropTypes.shape({
    name: PropTypes.string.isRequired,
    imageUrl: PropTypes.string.isRequired,
  }).isRequired,
}

export default Card
