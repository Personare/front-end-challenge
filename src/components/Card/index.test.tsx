import React from "react";
import { shallow } from "enzyme";
import Card from ".";
import { ICard } from "../../config/interfaces";

const CARD: ICard = {
  name: "Name Test",
  image: "Image Test",
  show: false,
};

const CONTEXT = {
  state: {
    data: {
      imagesUrl: "https://images.com/",
    },
  },
};

describe("Card", () => {
  beforeEach(() => {
    jest.spyOn(React, "useContext").mockImplementation(() => CONTEXT);
  });

  it("should match the snapshot", () => {
    const wrapper = shallow(<Card card={CARD} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("should call the function when clicking on the component", () => {
    const onClickMock = jest.fn();

    const wrapper = shallow(<Card card={CARD} onClick={onClickMock} />);
    wrapper.find("img").first().simulate("click");

    expect(onClickMock).toHaveBeenCalledWith(CARD.name);
  });
});
