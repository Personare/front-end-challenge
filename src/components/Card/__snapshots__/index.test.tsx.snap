// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match the snapshot 1`] = `
<ReactCardFlip
  cardStyles={
    Object {
      "back": Object {},
      "front": Object {},
    }
  }
  cardZIndex="auto"
  containerStyle={Object {}}
  flipDirection="horizontal"
  flipSpeedBackToFront={0.6}
  flipSpeedFrontToBack={0.6}
  infinite={false}
  isFlipped={true}
>
  <img
    alt="Name Test"
    onClick={[Function]}
    src="https://images.com/Image Test"
  />
  <img
    alt="Name Test"
    onClick={[Function]}
  />
</ReactCardFlip>
`;
